### ISSUE TITLE
Event length filter

### ISSUE THREAD

**Issue by epi-gene** [nanoporetech/bonito#30]
Would we be able to customize the event length filter on bonito. Feels like its is skipping many short reads in the run.

**Comment by iiSeymour** [comment]
Hey @epi-gene 

There is no explicit event length filter in bonito. Can you describe what you are seeing in more detail?

Are you saying when you are basecalling short reads you don't end up with a record in the fasta? What kind of sequence lengths are we talking?

Chris.

**Comment by epi-gene** [comment]
@iiSeymour Yes, We have some fragmentation products that are ~150 bp. Many of the them are being skipped by bonito, but can identified in guppy results. Bonito is also removing the sequencing barcodes aggressively.

**Comment by iiSeymour** [comment]
Okay, then I suspect it might be todo with the read trimming. Are you able to try the following patch which turns it off -

```diff
--- a/bonito/util.py
+++ b/bonito/util.py
@@ -81,7 +81,7 @@ def trim(signal, window_size=40, threshold_factor=3.0, min_elements=3):
 
 
 def preprocess(x, min_samples=1000):
-    start, end = trim(x)
+    start, end = 0, len(x)
     # REVISIT: we can potentially trim all the signal if this goes wrong
     if end - start < min_samples:
         start = 0
```

**Comment by epi-gene** [comment]
@iiSeymour The above patch made the sequences even shorter than earlier.

**Comment by iiSeymour** [comment]
Are you able to share a few examples for debugging - if so you can upload [here](https://nanoporetech.ent.box.com/upload-widget/view/zj4e5lmj9sjerlfgnxqc8tt4baisenc3/111261328980).

**Comment by epi-gene** [comment]
I have uploaded reads_original and reads_patch .fa files

**Comment by iiSeymour** [comment]
Can I get the `.fast5` also?

**Comment by epi-gene** [comment]
I have merged my individual .fast5's to run.fast5 - uploaded it to the portal

**Comment by epi-gene** [comment]
Also added a run.zip containing the individual .fast5's

**Comment by iiSeymour** [comment]
I think we have a few issues here @epi-gene see the unscaled raw read below, flicking through some the examples you uploaded, it is clear each read has failed to split correctly. Were you already aware Guppy was calling through multiple reads and are you happy with the resulting calls?

![1270db3c222858c99241c30a97342745c012a4262b967ba377d951b3](https://user-images.githubusercontent.com/4310904/80379371-ea433880-8895-11ea-8354-df0f10397de2.png)

Guppy is doing a better job of scaling the read and it's probably something I can improve but even with better scaling, I wouldn't want to emulate Guppy here.

#### Guppy (0.17 mean; 0.72 std). 
![08c8259a1c2f961af1049b3b545095c67a535f4dd742310cba81ae75](https://user-images.githubusercontent.com/4310904/80379182-a5b79d00-8895-11ea-831e-bff3e2b9a2dc.png)

#### Bonito  (0.60 mean; 5.77 std).

![31521960ba28d0fc229670874782e48ab62e1843d22f87a9df674093](https://user-images.githubusercontent.com/4310904/80379504-165eb980-8896-11ea-9ee1-3f3695ba7f26.png)

**Comment by epi-gene** [comment]
@iiSeymour Guppy was doing a better job in basecalling than bonito and I have recovered more strands by guppy than bonito. Yet, bonito's base accuracy was above. Do you recommend a tool for splitting a read manually ?

**Comment by iiSeymour** [comment]
I'm not convinced Guppy is doing something sensible here, do you have a reference sequence you can share?

**Comment by epi-gene** [comment]
@iiSeymour I have uploaded my Guppy results-fasta file (hac basecalled) and the ref_seq.fa file containing the strand we've sequenced.

**Comment by iiSeymour** [comment]
Sorry @epi-gene I've mapped your `guppy_results.fasta` to `ref_seq.fa` with minimap2 and the results confirm my suspicions, only 212/23036 reads actually map, and those that do are low coverage hits.

**Comment by epi-gene** [comment]
@iiSeymour  I agree with you. I have uploaded blastn matches for both bonito and guppy for reference. The number of matches to the total number of reads is quite low.

**Comment by iiSeymour** [comment]
I've had a look at the differences between the few good matches and found that a simpler scaling method does a better job here. Try the patch below; you should get a nice bump in both accuracy and the number of matches.  

```patch
 def get_raw_data(filename):
     """
     Get the raw signal and read id from the fast5 files
@@ -120,7 +133,7 @@ def get_raw_data(filename):
             scaling = channel_info['range'] / channel_info['digitisation']
             offset = int(channel_info['offset'])
             scaled = np.array(scaling * (raw + offset), dtype=np.float32)
-            yield read.read_id, preprocess(scaled)
+            yield read.read_id, (scaled - np.median(scaled)) / np.std(scaled)
 
 
 def window(data, size, stepsize=1, padded=False, axis=-1):
```

**Comment by epi-gene** [comment]
Initial data looks very promising ! I will let you know how the analysis goes on. Thanks !

**Comment by epi-gene** [comment]
Compared to the original ,both blastn and minimap2 yielded more significant matches and mapped reads respectively.

**Comment by iiSeymour** [comment]
@epi-gene I have just released version 0.1.5 which uses a better approach than my last suggestion resulting in more reads and higher accuracy once again.

**Comment by epi-gene** [comment]
@iiSeymour Great ! I'll check it out. Thanks.
